### DISCUSSION TITLE
foreach pattern matching

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#6689] [Language Ideas]
foreach could be extended to support pattern matching without creating a nested “if”.

Examples:
`foreach (var x when x is Some{Prop:1} in list)`
`foreach (int a when a is <10 and >1 in list)`

**Comment by DavidArno** [comment]
Since we can already write those as valid C# using:
```cs
foreach (var x in from l in list where l is Some{ Prop:1 } select l)
foreach (var a in from l in list where l is < 10 and > 1 select l)
```
It's unclear to me what advantage a new syntax would bring to the language. Other than being slightly shorter and not needing to introduce a variable, `l`, do you see any other benefits?

In addition, if the length is an issue, then the query syntax can be replaced with the fluent form:
```cs
foreach (var x in list.Where(l => l is Some{ Prop:1 }))
foreach (var a in list.Where(l => l is < 10 and > 1))
```
Here the length benefit is lost too so we only lose the extra variable. This seems a very small gain for yet another way of expressing the same thing.

**Reply by NN---** [reply]
The difference is introducing a lambda in LINQ expression.

I am not aware about LINQ rewriter to simple "if" plans in a near future.

Some parts of code or projects such as Roslyn for example avoid using lambdas where possible.

**Reply by DavidArno** [reply]
Ah, so you aren't so much after pattern matching in `foreach`, rather after being able to achieve this without the overhead of Linq. #2482 (Exploration: What would it take to achieve cost free Linq) might be of interest to you in that case.

**Reply by NN---** [reply]
There is a style and length difference between two examples, the later is slightly easier to read. 
Yes, the most notable difference is lambda overhead. 

`foreach (var x in from l in list where l is Some{ Prop:1 } select l)`

`foreach (var x when x is Some{ Prop:1 })`

**Comment by FaustVX** [comment]
Why you don't want to write
```C#
foreach (var x in list)
  if (x is Some { Prop: 1 })
  {
    // your code
  }
```

**Reply by NN---** [reply]
Most coding convention requires you putting braces so this becomes:

```cs
foreach (var x in list)
{
  if (x is Some { Prop: 1 })
  {
..
```

**Reply by Joe4evr** [reply]
![](https://i.imgur.com/jks6Qqr.png)

Apple's `goto` fail has demonstrated why that convention should be *more* adopted as the industry-wide standard.
